File: backend/internal/api/template_name_test.go
package api_test

import (
	"bytes"
	"encoding/json"
	"net/http"
	"os"
	"regexp"
	"strings"
	"testing"

	"github.com/stretchr/testify/require"
	"gopkg.in/yaml.v3"

	"kubeport/cmd/seed-demo/fixtures"
	"kubeport/internal/api"
	"kubeport/internal/config"
)

// Issue #369. A template's name is one path segment in every template route.
// A name with "/" was created with 201 and then answered 404 everywhere — no
// route could read, change, deploy or delete it — so creating one is refused,
// by the rule release names have.

// templateNamePatterns reads the pattern openapi.yaml gives the TemplateName
// path parameter and the one on CreateTemplateRequest.name.
func templateNamePatterns(t *testing.T) (param, body string) {
	t.Helper()
	b, err := os.ReadFile(specPath)
	require.NoError(t, err, "spec missing at %s", specPath)
	var doc struct {
		Components struct {
			Parameters map[string]struct {
				Schema struct {
					Pattern string `yaml:"pattern"`
				} `yaml:"schema"`
			} `yaml:"parameters"`
			Schemas map[string]struct {
				Properties map[string]struct {
					Pattern string `yaml:"pattern"`
				} `yaml:"properties"`
			} `yaml:"schemas"`
		} `yaml:"components"`
	}
	require.NoError(t, yaml.Unmarshal(b, &doc))
	return doc.Components.Parameters["TemplateName"].Schema.Pattern,
		doc.Components.Schemas["CreateTemplateRequest"].Properties["name"].Pattern
}

// Every name below gets the same answer from openapi.yaml's pattern as from
// the binding the API actually runs, so the document cannot promise a rule the
// server does not keep, or the other way round.
func TestOpenAPISpec_TemplateNamePatternIsTheBinding(t *testing.T) {
	param, body := templateNamePatterns(t)
	require.NotEmpty(t, param, "the TemplateName parameter has no pattern")
	require.Equal(t, param, body, "the path parameter and the create body must say the same")
	re := regexp.MustCompile(param)

	label := strings.Repeat("a", 63)
	for _, name := range []string{
		"web-app", "nightly-job", "app-with-config", "Web-App", "a", "0abc", "web.app", "web-",
		label, label + ".b", label + "a",
		"", "my/app", "my app", "..", ".web", "web.", "-web", "web_app", "웹앱", "web%2Fapp", "a?b", "a#b",
	} {
		specSays := re.MatchString(name)
		bindingSays := api.ValidateCreateTemplateName(name) == nil
		require.Equal(t, specSays, bindingSays, "%q: openapi.yaml says %v, the binding says %v", name, specSays, bindingSays)
	}
}

func TestCreateTemplate_RefusesANameNoRouteCanReach(t *testing.T) {
	admin := api.NewRouter(config.Config{}, api.Deps{Verifier: adminVerifier{}, Store: testStore(t)})
	for _, name := range []string{"my/app", "my app", "..", "-web", "web_app"} {
		w := do(t, admin, http.MethodPost, "/v1/templates", bytes.NewReader(createTemplateBody(name)))
		require.Equal(t, http.StatusBadRequest, w.Code, "%q: %s", name, w.Body.String())
		var p struct {
			Title string `json:"title"`
		}
		require.NoError(t, json.Unmarshal(w.Body.Bytes(), &p))
		require.Equal(t, "validation-error", p.Title, name)
	}
}

// The demo seeds its catalog under these names; a rule they broke would be a
// rule the live demo already breaks.
func TestTemplateName_TheDemoSeedPasses(t *testing.T) {
	param, _ := templateNamePatterns(t)
	re := regexp.MustCompile(param)
	require.NotEmpty(t, fixtures.All())
	for _, f := range fixtures.All() {
		require.NoError(t, api.ValidateCreateTemplateName(f.Name), f.Name)
		require.True(t, re.MatchString(f.Name), f.Name)
	}
}

File: frontend/lib/template-name.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import YAML from "yaml";

import { TEMPLATE_NAME_PATTERN, templateNameProblem } from "./template-name";

// The API's own document. The backend's spec test holds its pattern to the
// binding the server runs; this holds the form's copy to the document (#369).
// Tests run from frontend/ (pnpm -C frontend, and CI's working-directory); in
// the jsdom environment import.meta.url is not a file URL.
const spec = YAML.parse(
  readFileSync(resolve(process.cwd(), "../backend/api/openapi.yaml"), "utf8"),
) as {
  components: {
    parameters: Record<string, { schema?: { pattern?: string } }>;
    schemas: Record<string, { properties?: Record<string, { pattern?: string }> }>;
  };
};

describe("template name rule (#369)", () => {
  it("is the pattern openapi.yaml documents, on the path and on create", () => {
    expect(spec.components.parameters.TemplateName.schema?.pattern).toBe(TEMPLATE_NAME_PATTERN);
    expect(spec.components.schemas.CreateTemplateRequest.properties?.name.pattern).toBe(TEMPLATE_NAME_PATTERN);
  });

  it("passes the demo seed's names", () => {
    for (const name of ["web-app", "nightly-job", "app-with-config"]) {
      expect(templateNameProblem(name), name).toBeNull();
    }
  });

  it("refuses what no template route can reach", () => {
    for (const name of ["my/app", "my app", "..", ".web", "web.", "-web", "web_app", "a?b", "a#b", "웹앱"]) {
      expect(templateNameProblem(name), name).toBe("format");
    }
  });

  it("keeps what the API accepts, uppercase and dots included", () => {
    for (const name of ["Web-App", "0abc", "web.app", "a".repeat(63)]) {
      expect(templateNameProblem(name), name).toBeNull();
    }
    expect(templateNameProblem("a".repeat(64))).toBe("format");
  });

  it("tells an empty name apart", () => {
    expect(templateNameProblem("")).toBe("empty");
    expect(templateNameProblem("   ")).toBe("empty");
  });
});
